### ISSUE TITLE
Problem when compiling the DLL and the perpetual Attrition crashes

### ISSUE THREAD

**Issue by JFDaly** [LoneGazebo/Community-Patch-DLL#4001]
For the last few versions, I've been getting some errors when compiling the DLL, leading to its failure. I haven't changed my methods - download the repository, drop into Visual Studio 2013\Projects, load with Visual Studio, Build - but maybe I need to be doing something different now?

The output reads as follows:

![Alt Text](https://i.imgur.com/yhGSqTz.png)

I've finally re-added the Attrition component to my JFDLC compendium mod, and its still presenting with regular crashes. Disabled, JFDLC runs fine, enabled, almost guaranteed to crash. There's next to no Lua with Attrition (and disabling that doesn't fix the issue). It's hard to give you any concrete info without hooking up the DLL and trying to get some kind of callback, but perhaps its just a lost cause and I need to revert the system to Lua, or ditch it entirely.

Either way, any insight into the compiler issue would help and be appreciated.

**Comment by LoneGazebo** [comment]
Hmm. What specific DLL elements are you enabling or using for attrition?

**Comment by LoneGazebo** [comment]
If those problems are from last night's release, there was a merge error which threw a bunch of head <<< tags into the DLL. I just uploaded a merge fix.

**Comment by JFDaly** [comment]
Alright, I'll try redownloading.

For Attrition, its the LocationUnitFreePromotion tag (for Terrains/Features) that I'm using to give out Promotions that do damage. Also, the SpawnLocationUnitFreePromotion tag to give promotions with NegatesPromotion that obsoletes the ones from LocationUnitFreePromotion. I'll try disabling the latter and see if its spefically one over the other.

**Comment by JFDaly** [comment]
Yeah, redownloading allowed it to compile. I'll see what I can come up with for Attrition with it hooked up then, thank.

**Comment by LoneGazebo** [comment]
Cool, good to hear. If you want to test the change for your 'yieldmodifierfromdevelopment' request, it's an easy fix: go to line 25513 in CvCity.cpp.

Replace:

`int iStrengthMod = 0;

	// Player-wide strength mod (Policies, etc.)
	iStrengthMod += GET_PLAYER(getOwner()).GetCityStrengthMod();

	int iCrime = GetYieldModifierFromCrime(YIELD_JFD_CRIME);
	int iDevelopment = GetYieldModifierFromDevelopment(YIELD_JFD_CRIME);

	iStrengthMod += (iCrime + iDevelopment);`

with

`	iStrengthValue += GetYieldModifierFromDevelopment(YIELD_JFD_CRIME);

	int iStrengthMod = 0;

	// Player-wide strength mod (Policies, etc.)
	iStrengthMod += GET_PLAYER(getOwner()).GetCityStrengthMod();

	int iCrime = GetYieldModifierFromCrime(YIELD_JFD_CRIME);

	iStrengthMod += iCrime;`

**Comment by JFDaly** [comment]
Hm, is there somewhere else that adds it as raw Defense then? If I'm reading that right, that code'll just remove it as a modifier?

**Comment by LoneGazebo** [comment]
iStrengthValue += GetYieldModifierFromDevelopment(YIELD_JFD_CRIME);

that's the 'raw' value being added in.

**Comment by JFDaly** [comment]
Ah, gotcha.

**Comment by JFDaly** [comment]
Yeah, same situation as before. Can't trace anything back.

![Alt Text](https://i.imgur.com/1TfZnMR.png)

All that's running for Attrition is the (Location/Spawn) promotions from terrain/features part. The 'immunity' promotions that negate those promotions and all Lua is disabled.

**Comment by LoneGazebo** [comment]
If you can’t trace that usually means a lua crash. Are you attaching the dll to civ at runtime and letting it crash that way?

**Comment by JFDaly** [comment]
Hm, hooking it right after mods have configured right?

I really don't know at this point. There is no lua running with Attrition, yet, taking Attrition out and into its own mod seems not to trigger any crashes. It must be some weird interaction with the rest of JFDLC which on its own doesn't cause any issues. I think I'll just have to revert to handling Attrition via Lua.

**Comment by LoneGazebo** [comment]
Is it crashing, or is it hanging endlessly?

**Comment by JFDaly** [comment]
Crashes to desktop.

**Comment by LoneGazebo** [comment]
Okay. What function is attrition called by/in?

**Comment by LoneGazebo** [comment]
Oh, I see it. You could drop a breakpoint in DoAttrition in CvUnit.cpp, line 7903. Do you know how to do that?

**Comment by JFDaly** [comment]
Right-click the line > Breakpoint > Insert Breakpoint ? What would that do?

**Comment by LoneGazebo** [comment]
Yeah - so what you can do is you can watch the attrition code happen and walk through it line by line. Depending on where you set the breakpoint, you can spot check the code to see where, if anywhere in that function, the code is breaking.

**Comment by JFDaly** [comment]
Alright, will try that, thanks.

**Comment by LoneGazebo** [comment]
You can use the F keys (I think F9-11) to walk the code one step at a time after your break point.
